lib/spack/llnl/util/tty/log.py

```diff
@@ -89,180 +109,233 @@ def __enter__(self):
+            os.close(self._saved_stderr)
+        else:
+            sys.stdout = self._saved_stdout
+            sys.stderr = self._saved_stderr
+
+        # join the daemon process. The daemon will quit automatically
+        # when the write pipe is closed; we just wait for it here.
+        self.process.join()
+
+        # restore old color and debug settings
+        tty.color._force_color = self._saved_color
+        tty._debug = self._saved_debug
+
+        self._active = False  # safe to enter again
+
+    @contextmanager
+    def force_echo(self):
+        """Context manager to force local echo, even if echo is off."""
+        if not self._active:
+            raise RuntimeError(
+                "Can't call force_echo() outside log_output region!")
+
+        # This uses the xon/xoff to highlight regions to be echoed in the
+        # output. We us these control characters rather than, say, a
+        # separate pipe, because they're in-band and assured to appear
+        # exactly before and after the text we want to echo.
+        sys.stdout.write(xon)
+        sys.stdout.flush()
+        yield
+        sys.stdout.write(xoff)
+        sys.stdout.flush()
+
+    def _writer_daemon(self, stdin):
+        """Daemon that writes output to the log file and stdout."""
         # Use line buffering (3rd param = 1) since Python 3 has a bug
         # that prevents unbuffered text I/O.
-        read_file = os.fdopen(read, 'r', 1)
+        in_pipe = os.fdopen(self.read_fd, 'r', 1)
+        os.close(self.write_fd)
+
+        echo = self.echo        # initial echo setting, user-controllable
+        force_echo = False      # parent can force echo for certain output
 
         with open(self.filename, 'w') as log_file:
-            with keyboard_input(input_stream):
+            with keyboard_input(stdin):
                 while True:
-                    # Without the last parameter (timeout) select will wait
-                    # until at least one of the two streams are ready. This
-                    # may cause the function to hang.
-                    rlist, _, _ = select.select(
-                        [read_file, input_stream], [], [], 0
-                    )
+                    # Without the last parameter (timeout) select will
+                    # wait until at least one of the two streams are
+                    # ready. This may cause the function to hang.
+                    rlist, _, xlist = select.select(
+                        [in_pipe, stdin], [], [], 0)
 
                     # Allow user to toggle echo with 'v' key.
                     # Currently ignores other chars.
-                    if input_stream in rlist:
-                        if input_stream.read(1) == 'v':
-                            self.echo = not self.echo
+                    if stdin in rlist:
+                        if stdin.read(1) == 'v':
+                            echo = not echo
```

Comment:
I am currently not able to get this part working. I tested the PR on:
```console
$ uname -a
Linux nuvolari 4.4.0-87-generic #110~14.04.1-Ubuntu SMP Tue Jul 18 14:51:32 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
using different versions of python for spack.